### PR TITLE
feat: [indexer][v25.x] added test cases, token liquidity to token_swapped event, clean-up and fly-by fixes

### PR DESCRIPTION
**Pull request by cryptomatictrader**
This is the same changes as [PR8542](https://github.com/osmosis-labs/osmosis/pull/8542) just that this PR applies to v25.x branch vs [PR8542](https://github.com/osmosis-labs/osmosis/pull/8542) applies to main branch 

This PR contains the following changes. 

1. It addresses the requirement that the attribute `reserves` has to be provided in Dexscreener endpoint, though not mandatory. `reservers` refer to the pooled amount of each asset after a swap event has occurred. In this commit, pool id in the token_swapped event is retrieved, use GetTotalLiquidity to obtain the pool token liquidity, and then the token liquidity data is stored inside the attribute map of the event. Tested in v25.x and to be tested in sdk v50. Thanks.

2. It adds test cases to `pair` functions

3. It adds test cases to `AddTotalLiquidity` & `AdjustTokenInAmountBySpreadFactor` functions in `indexer_streaming_service.go`

4. It removes unused struct and functions 

5. Core code already running indexer staging and production. Main purposes of this branch is to add test cases to safeguard our work.

### CODE REVIEW

File: x/txfees/keeper/feedecorator.go [hunk 1]
```diff
@@ -62,7 +62,8 @@ func (mfd MempoolFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate b
 	}
 
 	// TODO: In v26, move this code to the correct spot, and make it CheckTX only.
-	if ctx.BlockHeight() >= 17004043 {
+	bh := ctx.BlockHeight()
+	if bh <= 16841115 || bh >= 17004043 {
```
Comment (p0mvn) [inline x/txfees/keeper/feedecorator.go:66]: Could we separate this into a separate PR so that this change is easier to track for historical purposes with its own PR description?

File: ingest/indexer/service/indexer_streaming_service.go [hunk 2]
```diff
@@ -240,6 +245,43 @@ func (s *indexerStreamingService) publishTxn(ctx context.Context, req types.Requ
 	return s.client.PublishTransaction(sdkCtx, txn)
 }
 
+// addTokenLiquidity adds the token liquidity to the event.
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service.go:248]:
```suggestion
// AddTokenLiquidity adds the token liquidity to the event.
```

File: ingest/indexer/service/indexer_streaming_service.go [hunk 3]
```diff
@@ -240,6 +245,43 @@ func (s *indexerStreamingService) publishTxn(ctx context.Context, req types.Requ
...
+		}
+	}
+	if poolIdStr == "" {
+		return errors.New("pool id not found")
+	}
+	poolId, err := strconv.ParseInt(poolIdStr, 10, 64)
+	if err != nil {
+		return err
+	}
+	coins, err := s.keepers.PoolManagerKeeper.GetTotalPoolLiquidity(sdkCtx, uint64(poolId))
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service.go:271]: nit: Is there a way to parse to uint directly?

File: ingest/indexer/service/indexer_streaming_service.go [hunk 4]
```diff
@@ -214,11 +214,16 @@ func (s *indexerStreamingService) publishTxn(ctx context.Context, req types.Requ
 	events := res.GetEvents()
 	var includedEvents []domain.EventWrapper
 	for i, event := range events {
-		err := s.adjustTokenInAmountBySpreadFactor(ctx, &event)
+		err := s.AdjustTokenInAmountBySpreadFactor(ctx, &event)
 		if err != nil {
 			s.logger.Error("Error adjusting amount by spread factor", "error", err)
 			continue
 		}
+		err = s.AddTokenLiquidity(ctx, &event)
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service.go:222]: Could we make sure that we make a deep copy of the event?

That is, we do not modify the event in the `res.GetEvents`.

I'm not sure how these are used in the caller but events are state-breaking, and I'm worried that if this modification is propagated to the SDK, we might have a consensus failure.

File: ingest/indexer/service/indexer_streaming_service.go [hunk 5]
```diff
@@ -240,6 +245,43 @@ func (s *indexerStreamingService) publishTxn(ctx context.Context, req types.Requ
 	return s.client.PublishTransaction(sdkCtx, txn)
 }
 
+// addTokenLiquidity adds the token liquidity to the event.
+// It refers to the pooled amount of each asset after a swap event has occurred.
+func (s *indexerStreamingService) AddTokenLiquidity(ctx context.Context, event *types.Event) error {
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service.go:250]: I think that it would be good to return a new deep copy of the event rather than mutating the given one for the state-breaking nature of events and possible edge cases around it that we can derisk with a deep copy.

File: ingest/indexer/service/indexer_streaming_service.go [hunk 6]
```diff
@@ -132,7 +132,7 @@ func (s *indexerStreamingService) publishBlock(ctx context.Context, req types.Re
 // as the amount in the event is the amount AFTER the spread factor is applied.
 // therefore, we need to adjust the amount by the spread factor to get the amount BEFORE the spread factor is applied.
 // NOTE: This applies to CL pools only
-func (s *indexerStreamingService) adjustTokenInAmountBySpreadFactor(ctx context.Context, event *types.Event) error {
+func (s *indexerStreamingService) AdjustTokenInAmountBySpreadFactor(ctx context.Context, event *types.Event) error {
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service.go:135]: Why do we export this method?

If only for tests, we can export it via `export_test.go` pattern (please see other file examples).

Keeping these unexported is helpful for the information-hiding principle, reducing the cognitive overhead for developers from having too many APIs exposed.

File: ingest/indexer/service/indexer_streaming_service_test.go [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+		})
+	}
+}
+
+func (s *IndexerServiceTestSuite) TestAddTokenLiquidity() {
+	testCases := []struct {
+		name                string
+		eventType           string
+		havePoolIDAttribute bool
+		useCLPoolID         bool
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service_test.go:249]: Please consider adding comments about what these fields control

File: ingest/indexer/service/indexer_streaming_service_test.go [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+	for _, tc := range testCases {
+		s.Run(tc.name, func() {
+			s.Setup()
+
+			// This test suite is to test the AddTokenLiquidity method in the indexer streaming service.
+			// where it applies to: token_swapped event only, i.e. gammtypes.TypeEvtTokenSwapped
+			// It then looks for the pool_id attribute (concentratedliquiditytypes.AttributeKeyPoolId) in the
+			// event attribute map.  With the pool_id, it then fetches the pool liquidity thru
+			// keepers.PoolManagerKeeper.GetTotalPoolLiquidity function. The pool liquidity data is then appended
+			// to the event attribute map with the key "liquidity_{denom}", value being the pool liquidity.
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service_test.go:295]: Nice!

File: ingest/indexer/service/indexer_streaming_service_test.go [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+					CFMMPools:         cfmmPools,
+					CosmWasmPools:     cosmWasmPools,
+				},
+			}
+
+			// Initialize a block process strategy manager
+			blockProcessStrategyManager := commondomain.NewBlockProcessStrategyManager()
+
+			// Initialize a concentrated pool with spread factor
+			sf, _ := math.LegacyNewDecFromStr(defaultSpreadFactor)
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service_test.go:340]: nit: let's use `osmomath.NewDecFromStr` for consistency with the rest of the codebase

File: ingest/indexer/service/indexer_streaming_service_test.go [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+			}
+			if tc.expectedLiquidity {
+				s.Require().True(foundKey0)
+				s.Require().True(foundKey1)
+			} else {
+				s.Require().False(foundKey0)
+				s.Require().False(foundKey1)
+			}
+
+		})
```
Comment (p0mvn) [inline ingest/indexer/service/indexer_streaming_service_test.go:421]: I wonder if we could encapsulate some of the verbose boilerplate into test helper methods so that the core test logic is more simple and readable.

I currently feel like there is a lot going on in the test

**Review by p0mvn — APPROVED** [review]
Nice work!

Left some comments but should be non-blocking the merge after addressed. I think that we should separate the fee decorator PR though for historical tracking since it affects core chain flows.

Just confirming - has this also been tested / dry run in the dev environment? If not, let's please do that since the changes are quite involved 🙏

File: x/txfees/keeper/feedecorator.go (hunk 1 above)
Comment (cryptomatictrader) [reply x/txfees/keeper/feedecorator.go:66]: Make sense. Created separate PR for forward porting it to v26, and let mergify to kick in and merge the changes into v25.

PR: https://github.com/osmosis-labs/osmosis/pull/8629

File: ingest/indexer/service/indexer_streaming_service.go (hunk 2 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service.go:248]: Will keep it as small letter as this will not be exported by using export_test.go

File: ingest/indexer/service/indexer_streaming_service.go (hunk 3 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service.go:271]: Got it 👍 good one 👍

File: ingest/indexer/service/indexer_streaming_service.go (hunk 6 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service.go]: Good suggestion. Done.

File: ingest/indexer/service/indexer_streaming_service_test.go (hunk 7 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service_test.go]: Good point. Done.

File: ingest/indexer/service/indexer_streaming_service_test.go (hunk 9 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service_test.go:341]: Done 👍

File: ingest/indexer/service/indexer_streaming_service.go (hunk 4 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service.go]: VERY GOOD POINT! 
Done.

File: ingest/indexer/service/indexer_streaming_service.go (hunk 5 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service.go]: Done as part of the deepCloneEvent function

File: ingest/indexer/service/indexer_streaming_service_test.go (hunk 10 above)
Comment (cryptomatictrader) [reply ingest/indexer/service/indexer_streaming_service_test.go:422]: Good point. Done by externalize such block of the code into a function named `checkIfLiquidityAttributeExists`

**Comment by cryptomatictrader** [comment]
> Nice work!
> 
> Left some comments but should be non-blocking the merge after addressed. I think that we should separate the fee decorator PR though for historical tracking since it affects core chain flows.
> 
> Just confirming - has this also been tested / dry run in the dev environment? If not, let's please do that since the changes are quite involved 🙏

Thanks Roman! 
Recommendations well taken! And yes - the tests have been tested in dev.

**Comment by cryptomatictrader** [comment]
Finally merging into v25.x ! 
Will apply the same recommendations in v26 main.
